yapapi/services.py

```diff
@@ -132,31 +128,39 @@ def __init__(self, cluster: "Cluster", ctx: WorkContext, network_node: Optional[
         # Tuple of `None`'s means that the transition was not caused by an exception.
         self._exc_info: ExcInfo = (None, None, None)
 
+        self.__service_instance = ServiceInstance(self)
+
         # TODO: maybe transition due to a control signal should also set this? To distinguish
         # stopping the service externally (e.g., via `cluster.stop()`) from internal transition
         # (e.g., via returning from `Service.run()`).
 
     @property
-    def cluster(self) -> "Cluster":
+    def cluster(self) -> Optional["Cluster"]:
         """Return the Cluster to which this service instance belongs."""
         return self._cluster
 
     @property
-    def id(self) -> str:
+    def id(self) -> Optional[str]:
```

Comment:
hmm... if we're claiming in the comment below that it's "Guaranteed to be unique within a :class:`yapapi.services.Cluster`."  then it should not be `Optional`... I'd make it non-optional and just stop binding it with the WorkContext's id and just replace it with a uuid4() ...

Reply:
937b39c